[SPARK-1517] Refactor release scripts to facilitate nightly publishing

pwendell opened the pull request:
This update contains some code changes to the release scripts that allow easier nightly publishing. I've been using these new scripts on Jenkins for cutting and publishing nightly snapshots for the last month or so, and it has been going well. I'd like to get them merged back upstream so this can be maintained by the community.

The main changes are:
1. Separates the release tagging from various build possibilities for an already tagged release (`release-tag.sh` and `release-build.sh`).
2. Allow for injecting credentials through the environment, including GPG keys. This is then paired with secure key injection in Jenkins.
3. Support for copying build results to a remote directory, and also "rotating" results, e.g. the ability to keep the last N copies of binary or doc builds.

I'm happy if anyone wants to take a look at this - it's not user facing but an internal utility used for generating releases.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #37307 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37307/consoleFull) for   PR 7411 at commit [`d28c240`](https://github.com/apache/spark/commit/d28c24003e09ba7290ec6d7bef1d59d8406f9915).

SparkQA commented:
[Test build #37307 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37307/console) for   PR 7411 at commit [`d28c240`](https://github.com/apache/spark/commit/d28c24003e09ba7290ec6d7bef1d59d8406f9915).
 * This patch **fails RAT tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class Exchange(newPartitioning: Partitioning, child: SparkPlan) extends UnaryNode `

AmplabJenkins commented:
Merged build finished. Test FAILed.

JoshRosen commented:
Aw, looks like this conflicted with some recent Maven changes.

pwendell commented:
Yeah I need to update this - but it's off the critical path for the release
so it's on hold.

On Sat, Aug 1, 2015 at 10:59 AM, Josh Rosen <notifications@github.com>
wrote:

> Aw, looks like this conflicted with some recent Maven changes.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/spark/pull/7411#issuecomment-126942487>.
>

pwendell committed:
[SPARK-1517] Refactor release scripts to facilitate nightly publishing

This update contains some code changes to the release scripts that allow easier nightly publishing. I've been using these new scripts on Jenkins for cutting and publishing nightly snapshots for the last month or so, and it has been going well. I'd like to get them merged back upstream so this can be maintained by the community.

The main changes are:
1. Separates the release tagging from various build possibilities for an already tagged release (`release-tag.sh` and `release-build.sh`).
2. Allow for injecting credentials through the environment, including GPG keys. This is then paired with secure key injection in Jenkins.
3. Support for copying build results to a remote directory, and also "rotating" results, e.g. the ability to keep the last N copies of binary or doc builds.

I'm happy if anyone wants to take a look at this - it's not user facing but an internal utility used for generating releases.

pwendell commented:
Jenkins, test this please.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

ryan-williams commented on dev/create-release/release-build.sh:
```diff
@@ -0,0 +1,320 @@
+
+  ./dev/change-scala-version.sh 2.11
+
+  build/mvn -DzincPort=$ZINC_PORT -Dmaven.repo.local=$tmp_repo -Dscala-2.11 \
```
you can drop these `-DzincPort=$ZINC_PORT`s as of #7944 right?

pwendell replied:
Yeah true, thanks for pointing that out.

On Mon, Aug 10, 2015 at 2:05 PM, Ryan Williams <notifications@github.com>
wrote:

> In dev/create-release/release-build.sh
> <https://github.com/apache/spark/pull/7411#discussion_r36684398>:
>
> > +    -H "Content-Type:application/xml" -v \
> > +    $NEXUS_ROOT/profiles/$NEXUS_PROFILE/start)
> > +  staged_repo_id=$(echo $out | sed -e "s/.*\(orgapachespark-[0-9]\{4\}\).*/\1/")
> > +  echo "Created Nexus staging repository: $staged_repo_id"
> > +
> > +  tmp_repo=$(mktemp -d spark-repo-XXXXX)
> > +
> > +  # Generate random point for Zinc
> > +  export ZINC_PORT=$(python -S -c "import random; print random.randrange(3030,4030)")
> > +
> > +  build/mvn -DzincPort=$ZINC_PORT -Dmaven.repo.local=$tmp_repo -DskipTests $PUBLISH_PROFILES \
> > +    -Phive-thriftserver clean install
> > +
> > +  ./dev/change-scala-version.sh 2.11
> > +
> > +  build/mvn -DzincPort=$ZINC_PORT -Dmaven.repo.local=$tmp_repo -Dscala-2.11 \
>
> you can drop these -DzincPort=$ZINC_PORTs as of #7944
> <https://github.com/apache/spark/pull/7944> right?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/spark/pull/7411/files#r36684398>.
>

pwendell replied:
Actually since this is still used to build older versions, I think we'll
need to keep it for now, but we can eventually drop it.

On Mon, Aug 10, 2015 at 2:07 PM, Patrick Wendell <patrick@databricks.com>
wrote:

> Yeah true, thanks for pointing that out.
>
> On Mon, Aug 10, 2015 at 2:05 PM, Ryan Williams <notifications@github.com>
> wrote:
>
>> In dev/create-release/release-build.sh
>> <https://github.com/apache/spark/pull/7411#discussion_r36684398>:
>>
>> > +    -H "Content-Type:application/xml" -v \
>> > +    $NEXUS_ROOT/profiles/$NEXUS_PROFILE/start)
>> > +  staged_repo_id=$(echo $out | sed -e "s/.*\(orgapachespark-[0-9]\{4\}\).*/\1/")
>> > +  echo "Created Nexus staging repository: $staged_repo_id"
>> > +
>> > +  tmp_repo=$(mktemp -d spark-repo-XXXXX)
>> > +
>> > +  # Generate random point for Zinc
>> > +  export ZINC_PORT=$(python -S -c "import random; print random.randrange(3030,4030)")
>> > +
>> > +  build/mvn -DzincPort=$ZINC_PORT -Dmaven.repo.local=$tmp_repo -DskipTests $PUBLISH_PROFILES \
>> > +    -Phive-thriftserver clean install
>> > +
>> > +  ./dev/change-scala-version.sh 2.11
>> > +
>> > +  build/mvn -DzincPort=$ZINC_PORT -Dmaven.repo.local=$tmp_repo -Dscala-2.11 \
>>
>> you can drop these -DzincPort=$ZINC_PORTs as of #7944
>> <https://github.com/apache/spark/pull/7944> right?
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/apache/spark/pull/7411/files#r36684398>.
>>
>
>

SparkQA commented:
[Test build #40321 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/40321/consoleFull) for   PR 7411 at commit [`233ce85`](https://github.com/apache/spark/commit/233ce856914d75b55321527c220cdb39622d4518).

SparkQA commented:
[Test build #40321 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/40321/console) for   PR 7411 at commit [`233ce85`](https://github.com/apache/spark/commit/233ce856914d75b55321527c220cdb39622d4518).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

pwendell committed:
Moving maven build command to a variable

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #40327 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/40327/consoleFull) for   PR 7411 at commit [`74f9beb`](https://github.com/apache/spark/commit/74f9bebbaf6da0ea9ebd0216cabeaaeb303b6bd1).

SparkQA commented:
[Test build #40327 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/40327/console) for   PR 7411 at commit [`74f9beb`](https://github.com/apache/spark/commit/74f9bebbaf6da0ea9ebd0216cabeaaeb303b6bd1).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

pwendell commented:
Jenkins, retest this please.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #40397 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/40397/consoleFull) for   PR 7411 at commit [`74f9beb`](https://github.com/apache/spark/commit/74f9bebbaf6da0ea9ebd0216cabeaaeb303b6bd1).

SparkQA commented:
[Test build #40397 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/40397/console) for   PR 7411 at commit [`74f9beb`](https://github.com/apache/spark/commit/74f9bebbaf6da0ea9ebd0216cabeaaeb303b6bd1).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

pwendell commented:
Jenkins, retest this please.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

AmplabJenkins commented:
Merged build finished. Test FAILed.

pwendell commented:
Jenkins, test this please.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #40463 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/40463/consoleFull) for   PR 7411 at commit [`74f9beb`](https://github.com/apache/spark/commit/74f9bebbaf6da0ea9ebd0216cabeaaeb303b6bd1).

pwendell commented:
test

SparkQA commented:
[Test build #40463 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/40463/console) for   PR 7411 at commit [`74f9beb`](https://github.com/apache/spark/commit/74f9bebbaf6da0ea9ebd0216cabeaaeb303b6bd1).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

JoshRosen commented:
Jenkins, retest this please.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #40500 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/40500/consoleFull) for   PR 7411 at commit [`74f9beb`](https://github.com/apache/spark/commit/74f9bebbaf6da0ea9ebd0216cabeaaeb303b6bd1).

SparkQA commented:
[Test build #40500 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/40500/console) for   PR 7411 at commit [`74f9beb`](https://github.com/apache/spark/commit/74f9bebbaf6da0ea9ebd0216cabeaaeb303b6bd1).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

pwendell commented:
Okay will merge this - I've been keeping things in a separate repo and it's much better to have it in the upstream in case others want to modify it.
